[Java] `ThreadSafeFory` does not appear to be thread-safe when registering classes with `createSerializer` = true

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

0.11.0

### Component(s)

Java

### Minimal reproduce step

```groovy
(0..1000).collect {
    new Thread(() -> {
        Fory.builder()
            .withLanguage(Language.JAVA)
            .requireClassRegistration(true)
            .withAsyncCompilation(true)
            .withCompatibleMode(CompatibleMode.COMPATIBLE)
            .buildThreadSafeForyPool(2, 4, 1, TimeUnit.MINUTES)
            .register(MyClass.class, true) // register custom class and create serializer
        }
    )
}.collect {
    it.start()
    it
}.each {
    it.join()
}
```

### What did you expect to see?

No errors

### What did you see instead?

```
java.lang.RuntimeException: Create sequential serializer failed, 
class: class MyClass
	at org.apache.fory.serializer.CodegenSerializer.loadCodegenSerializer(CodegenSerializer.java:52)
	at org.apache.fory.resolver.ClassResolver.lambda$getObjectSerializerClass$4(ClassResolver.java:1138)
	at org.apache.fory.builder.JITContext.lambda$registerSerializerJITCallback$0(JITContext.java:91)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1095)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:619)
	at java.base/java.lang.Thread.run(Thread.java:1447)
Caused by: java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1230)
	at org.apache.fory.resolver.ClassResolver.getClassDef(ClassResolver.java:1747)
	at org.apache.fory.builder.ObjectCodecBuilder.lambda$new$0(ObjectCodecBuilder.java:97)
	at org.apache.fory.builder.JITContext.asyncVisitFory(JITContext.java:159)
	at org.apache.fory.builder.BaseObjectCodecBuilder.fory(BaseObjectCodecBuilder.java:225)
	at org.apache.fory.builder.ObjectCodecBuilder.<init>(ObjectCodecBuilder.java:94)
	at org.apache.fory.builder.CodecUtils.loadOrGenObjectCodecClass(CodecUtils.java:43)
	at org.apache.fory.serializer.CodegenSerializer.loadCodegenSerializer(CodegenSerializer.java:49)
	... 5 more
```

### Anything Else?

The culprit seems to be line 242 in `ClassResolver.java`:

`private final Map<Class<?>, ClassDef> classDefMap = new HashMap<>();`

Changing that to `ConcurrentHashMap` would probably solve the issue, but I'm not sure if that would break any other contracts.

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
+++ b/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
@@ -1394,6 +1394,15 @@ public class ClassResolver implements TypeResolver {
   }
 
   private void createSerializerAhead(Class<?> cls) {
+    try {
+      fory.getJITContext().lock();
+      createSerializer0(cls);
+    } finally {
+      fory.getJITContext().unlock();
+    }
+  }
+
+  private void createSerializer0(Class<?> cls) {
     ClassInfo classInfo = getClassInfo(cls);
     ClassInfo deserializationClassInfo;
     if (metaContextShareEnabled && needToWriteClassDef(classInfo.serializer)) {

--- a/java/fory-core/src/test/java/org/apache/fory/builder/JITContextTest.java
+++ b/java/fory-core/src/test/java/org/apache/fory/builder/JITContextTest.java
@@ -24,10 +24,14 @@ import static org.testng.Assert.assertTrue;
 
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Sets;
+import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.CountDownLatch;
 import lombok.Data;
 import org.apache.fory.Fory;
 import org.apache.fory.ForyTestBase;
+import org.apache.fory.ThreadSafeFory;
 import org.apache.fory.config.CompatibleMode;
 import org.apache.fory.config.Language;
 import org.apache.fory.logging.Logger;
@@ -165,6 +169,41 @@ public class JITContextTest extends ForyTestBase {
     assertTrue(ReflectionUtils.getObjectFieldValue(serializer, "serializer") instanceof Generated);
     assertTrue(ReflectionUtils.getObjectFieldValue(serializer, "serializer1") instanceof Generated);
     serDeCheck(fory, o);
+  }
+
+  @Test(timeOut = 60000)
+  public void testThreadSafetyWithManyForyInstances() throws Exception {
+    final int threadCount = 1000;
+    final CountDownLatch latch = new CountDownLatch(threadCount);
+    final List<Throwable> errors = Collections.synchronizedList(new ArrayList<>());
+
+    for (int i = 0; i < threadCount; i++) {
+      new Thread(
+              () -> {
+                try {
+                  ThreadSafeFory fory =
+                      Fory.builder()
+                          .withLanguage(Language.JAVA)
+                          .requireClassRegistration(true)
+                          .withAsyncCompilation(true)
+                          .withCompatibleMode(CompatibleMode.COMPATIBLE)
+                          .buildThreadSafeForyPool(2, 4, 1, java.util.concurrent.TimeUnit.MINUTES);
+                  fory.register(BeanB.class, true);
+                  // fory.register(BeanA.class, true);
+                } catch (Throwable t) {
+                  errors.add(t);
+                } finally {
+                  latch.countDown();
+                }
+              })
+          .start();
+    }
+    latch.await();
+    // print stack trace in errors
+    for (Throwable error : errors) {
+      error.printStackTrace();
+    }
+    assertTrue(errors.isEmpty(), "No exceptions should be thrown: " + errors);
   }
 
   @Data

